ocw/lib/openstack.py

```diff
@@ -0,0 +1,87 @@
+from datetime import datetime, timezone
+from typing import Dict
+from dateutil.parser import parse
+import openstack
+from webui.PCWConfig import PCWConfig
+from .provider import Provider
+
+
+class Openstack(Provider):
+    __instances: Dict[str, "Openstack"] = {}
+
+    def __init__(self, namespace: str):
+        super().__init__(namespace)
+        self.client()
+
+    def __new__(cls, vault_namespace: str):
+        if vault_namespace not in Openstack.__instances:
+            Openstack.__instances[vault_namespace] = self = object.__new__(cls)
+            self.__client = None
+        return Openstack.__instances[vault_namespace]
+
+    def client(self) -> None:
+        if self.__client is None:
+            self.__client = openstack.connect(
+                insecure=True,  # Trust the certificate
+                auth_url=self.get_data('auth_url'),
+                project_name=self.get_data('project_name'),
+                username=self.get_data('username'),
+                password=self.get_data('password'),
+                region_name=self.get_data('region_name'),
+                user_domain_name=self.get_data('user_domain_name'),
+                project_domain_name=self.get_data('project_domain_id'),
+            )
+        return self.__client
+
+    def cleanup_all(self) -> None:
+        self._cleanup_instances()
+        self._cleanup_images()
+        self._cleanup_keypairs()
+
+    def _cleanup_instances(self) -> None:
+        # Delete VM's & associated floating IP address(es)
+        servers = [vm for vm in self.client().compute.servers() if vm.startswith("openqa-vm-")]
+        self.log_dbg("Found {} servers", len(servers))
+        for server in servers:
+            if not self.is_outdated(parse(server.created_at)):
+                continue
+            if self.dry_run:
+                self.log_info("Instance termination {} skipped due to dry run mode", server.name)
+            else:
+                self.log_info("Deleting instance {}", server.name)
+                if not self.client().delete_server(
```

Comment:
According to the [documentation](https://docs.openstack.org/openstacksdk/latest/user/proxies/compute.html), this always returns `None`.

Following [some examples](https://docs.openstack.org/newton/user-guide/sdk-compute-apis.html#delete-server-api-v2) I couldn't find anything on how to detect errors here. Do you know a better way of doing this, while also being able to get errors?

Reply:
I'm using the `delete_server()` method of the `openstack.connection.Connection` class which returns a boolean value:

https://docs.openstack.org/openstacksdk/latest/user/connection.html#openstack.connection.Connection

I will add code to handle the OpenStackCloudException in every API call, though.

Reply:
Ahhh, I see. Thanks! Don't we need to catch `OpenStackCloudException` then?